# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/base/TextDirectiveCreator.cpp
# Commit: 5fe5a185035e
# Full Hash: 5fe5a185035e90b142f267c768f8b019256f25d9
# Author: Jan-Niklas Jaeschke <jjaschke@mozilla.com>
# Date: 2025-09-23 17:35 +0000
# Description:
#   Bug 1989891 - Text Fragments: Ensure context terms don't start/end with whitespace when reaching maximum length. r=smaug
# ==============================================================================

diff -r c84a239b905e -r 5fe5a185035e dom/base/TextDirectiveCreator.cpp
--- a/dom/base/TextDirectiveCreator.cpp	Tue Sep 23 17:01:42 2025 +0000
+++ b/dom/base/TextDirectiveCreator.cpp	Tue Sep 23 17:35:19 2025 +0000
@@ -302,12 +302,18 @@
   mPrefixContent =
       MOZ_TRY(TextDirectiveUtil::RangeContentAsString(prefixRange));
   if (mPrefixContent.Length() > kMaxContextTermLength) {
+    // cut off the prefix content at a word boundary near the max context term
+    // length to make sure the term does not start with whitespace.
+    auto [wordBegin, wordEnd] = intl::WordBreaker::FindWord(
+        mPrefixContent, mPrefixContent.Length() - kMaxContextTermLength);
+    while (nsContentUtils::IsHTMLWhitespace(mPrefixContent.CharAt(wordBegin))) {
+      ++wordBegin;
+    }
     TEXT_FRAGMENT_LOG(
         "Prefix term seems very long ({} chars), "
         "only considering the last {} chars.",
-        mPrefixContent.Length(), kMaxContextTermLength);
-    mPrefixContent = Substring(mPrefixContent,
-                               mPrefixContent.Length() - kMaxContextTermLength);
+        mPrefixContent.Length(), wordBegin);
+    mPrefixContent = Substring(mPrefixContent, wordBegin);
   }
   mPrefixFoldCaseContent = mPrefixContent;
   ToFoldedCase(mPrefixFoldCaseContent);
@@ -332,11 +338,19 @@
   mSuffixContent =
       MOZ_TRY(TextDirectiveUtil::RangeContentAsString(suffixRange));
   if (mSuffixContent.Length() > kMaxContextTermLength) {
+    // cut off the suffix content at a word boundary near the max context term
+    // length to make sure the term does not end with whitespace.
+    auto [wordBegin, wordEnd] =
+        intl::WordBreaker::FindWord(mSuffixContent, kMaxContextTermLength);
+    while (
+        nsContentUtils::IsHTMLWhitespace(mSuffixContent.CharAt(wordEnd - 1))) {
+      --wordEnd;
+    }
     TEXT_FRAGMENT_LOG(
         "Suffix term seems very long ({} chars), "
         "only considering the first {} chars.",
-        mSuffixContent.Length(), kMaxContextTermLength);
-    mSuffixContent = Substring(mSuffixContent, 0, kMaxContextTermLength);
+        mSuffixContent.Length(), wordEnd + 1);
+    mSuffixContent = Substring(mSuffixContent, 0, wordEnd + 1);
   }
   mSuffixFoldCaseContent = mSuffixContent;
   ToFoldedCase(mSuffixFoldCaseContent);